### DISCUSSION TITLE
Install IPSec policy with narrower traffic selector than negotiated?

### DISCUSSION THREAD

**Discussion by rionda** [strongswan/strongswan#923] [Q&A] [answered]
I'm using strongSwan to connect as a roadwarrior to a responder that does not allow me to negotiate anything different than `0.0.0.0/0` as traffic selector (which I have to specify in `remote_ts`). Is there any way to make strongSwan install an IPSec policy that uses a narrower (specified by me) traffic selector? The responder shouldn't care nor notice that I'm not fully using the negotiated traffic selector.

**Comment by tobiasbrunner** [comment, marked as answer]
Since you are probably using a virtual IP, you don't have to change the IPsec policies as they only match packets that have the virtual IP as source address. This is achieved by a route installed in table 220 that sets that IP as preferred source.

So to only tunnel selected traffic, you could disable automatic route installation via _charon.install_routes_ and manually install routes for the subnets you want to reach. You can use a custom updown or vici script (or a dispatch script if you use the NM plugin) that installs appropriate routes when the connection is established.

For instance, in an updown script this could be something like `ip route add <subnet> dev $PLUTO_INTERFACE src $PLUTO_MY_SOURCEIP table 220`.

**Reply by rionda** [reply]
Thanks for the reply. 
The solution does work, but it does not feel very elegant, in part because `charon.install_routes` is a "strongSwan-wide" setting, while maybe I want not to install routes for one connection (or even child), and I do want to install routes for another connection or child. I think that if `install_routes` was a setting of `<connection>.children.<child>` it would be more elegant (perhaps that can be added?)

But an even more elegant solution would allow to achieve "policy-based-only" IPSec, rather than a mix of policy- and route-based, with a `<connection>.chidren.<child>.override_remote_ts` settings that would only add the specified subnets to the policy.

The above are obviously just humble suggestions for potential features, the suggested solution works.

**Reply by rionda** [reply]
But maybe my second suggestion is moot because if you have a vIP you need at least one route that uses it if you want to send any traffic through the tunnel.

**Reply by tobiasbrunner** [reply]
Plugins can override traffic selectors before they are installed locally, so you could implement it that way if you like (routes would automatically get installed based on these narrowed traffic selectors). The other things are good suggestions but have currently no priority.

**Reply by rionda** [reply]
Thank you, @tobiasbrunner . BTW, amazing work and support.

**Reply by tobiasbrunner** [reply]
Please see [updown](https://wiki.strongswan.org/projects/strongswan/wiki/updown) and [vici](https://wiki.strongswan.org/projects/strongswan/wiki/Vici). So there are a lot of differences. A major one is that updown scripts are called for every combination of local and remote traffic selectors of a CHILD_SA, while the _child-updown_ vici event is only triggered once when a CHILD_SA is established (a simple example for a Python script that's kinda similar to a classic updown script can be found [here](https://github.com/strongswan/strongswan/blob/master/testing/tests/route-based/net2net-xfrmi-ike/hosts/sun/etc/updown.py)).
